### PR TITLE
chore: add integration test for provider

### PR DESCRIPTION
**Pull request by johnstcn**
This PR adds integration tests for the provider against an arbitrary version of Coder.

Example usage:
```
CODER_VERSION="v2.10.0-devel-3a9a7d199-amd64" go test -v ./integration
```

Testing methodology:
- Create an ephemeral Coder instance running in a container
- Perform initial setup
- Import a predefined template that creates a single file
- Create a workspace that populates the file with the actual output
- Compare the actual versus expected output

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -1,8 +1,11 @@
 module github.com/coder/terraform-provider-coder
 
-go 1.20
+go 1.21
+
+toolchain go1.22.3
```
Comment (johnstcn) [inline go.mod:5]: This was not a deliberate change and is likely due to importing `docker/docker`.

**Review by dannykopping — COMMENTED** [review]
Looking super slick, added a few notes

File: integration/integration_test.go [hunk 2]
```diff
@@ -141,24 +179,39 @@ func setup(ctx context.Context, t *testing.T) string {
...
+	// nolint:gosec // For testing only.
 	var (
 		testEmail    = "testing@coder.com"
 		testPassword = "InsecurePassw0rd!"
 		testUsername = "testing"
 	)
 
 	// Wait for container to come up
-	execContainer(ctx, t, ctr.ID, `until curl -s --fail http://localhost:3000/healthz; do sleep 1; done`)
+waitLoop:
```
Comment (dannykopping) [inline integration/integration_test.go:190]: I think `require.Eventually` might be more idiomatic?

File: integration/test-data-source/main.tf [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    coder = {
+      source = "coder/coder"
+    }
+    local = {
+      source = "hashicorp/local"
+    }
+  }
+}
+
+// TODO: test coder_external_auth and coder_git_auth
```
Comment (matifali) [inline integration/test-data-source/main.tf:12]: `coder_git_auth` could be skipped as it has been deprecated for a while.

File: integration/test-data-source/main.tf [hunk 4]
```diff
@@ -0,0 +1,65 @@
+terraform {
```
Comment (mtojek) [inline integration/test-data-source/main.tf:1]: nit: curious if we can use a file from the examples directory, or put it there.

**Review by mtojek — APPROVED** [review]
Nice testing improvement, so let's get it in.

File: integration/test-data-source/main.tf (hunk 4 above)
Comment (johnstcn) [reply integration/test-data-source/main.tf:1]: We could use some, but most have dependencies like gcloud credentials that we can't necessarily provide.

**Comment by johnstcn** [comment]
Will add to GitHub testing workflow in a separate PR.

**Pull request merged**